src/pyhf/tensor/tensorflow_backend.py

```diff
@@ -195,26 +186,20 @@ def reshape(self, tensor, newshape):
         return tf.reshape(tensor, newshape)
 
     def divide(self, tensor_in_1, tensor_in_2):
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
         return tf.divide(tensor_in_1, tensor_in_2)
 
     def log(self, tensor_in):
-        tensor_in = self.astensor(tensor_in)
         return tf.math.log(tensor_in)
 
     def exp(self, tensor_in):
-        tensor_in = self.astensor(tensor_in)
         return tf.exp(tensor_in)
 
     def stack(self, sequence, axis=0):
         return tf.stack(sequence, axis=axis)
 
     def where(self, mask, tensor_in_1, tensor_in_2):
-        mask = self.astensor(mask)
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
-        return mask * tensor_in_1 + (1 - mask) * tensor_in_2
+        mask = tf.cast(mask, tf.bool)
```

Comment:
why is there a cast here? The mask should be `bool` already.

Reply:
> why is there a cast here? The mask should be `bool` already.

I think that it is better to just cast it to make sure. There are times like the following where the mask seems like a valid `Bool`

https://github.com/diana-hep/pyhf/blob/c8a94f4d5a11edfdd5a28184153e1ef4b076e638/tests/test_tensor.py#L48-L50

but TensorFlow doesn't see it as such if the `tf.cast(mask, tf.bool)` is removed. So I think that it should just be used to ensure the right Tensor `dtype`.

Reply:
https://github.com/diana-hep/pyhf/blob/d9aeac8bd3b7bd384342ae2e0e7547b78714e66c/src/pyhf/tensor/tensorflow_backend.py#L124

That's because the default cast is to float. You should use `tb.astensor(tensor, dtype='bool')` for the mask input.

Reply:
tl;dr - the test is wrong.

Reply:
> That's because the default cast is to float

Yes, I wasn't confused as to the _why_ of the behavior, but as to the reasonableness of it. I will change the test, but if you were a user it would seem to me that `tb.astensor([1, 0, 1])` could be a valid way to indicate a Bool mask. So I would opt to include that cast there, which should be faster then the `self.astensor()` call that we were previously including.